File: ising.py
import numpy as np
import numba

# Initialize the lattice with spins based on initial up rate
@numba.njit(nopython=True, nogil=True)
def initialize_lattice(L: int, init_up_rate: float) -> tuple:
    lattice = np.where(np.random.random_sample((L, L)) < init_up_rate, 1, -1)
    magnetization = np.sum(lattice)
    energy = compute_total_energy(lattice, L)
    return lattice, magnetization, energy

# Compute the total energy of the lattice
@numba.njit(nopython=True, nogil=True)
def compute_total_energy(lattice: np.ndarray, L: int) -> float:
    kernel = np.array([[False,  True, False], 
                       [ True, False,  True], 
                       [False,  True, False]])

    padded = np.zeros((L + 2, L + 2), dtype=np.int32)
    padded[1:-1, 1:-1] = lattice

    # Manually wrap the borders to simulate periodic boundary conditions
    padded[:,0] = padded[:,-2]
    padded[:,-1] = padded[:,1]
    padded[0,:] = padded[-2,:]
    padded[-1,:] = padded[1,:]

    convolution = np.zeros((L, L), dtype=np.float64)

    for y in range(L):
        for x in range(L):
            convolution[y, x] = np.sum(padded[y:y + 3, x:x + 3] * kernel)

    return (-lattice * convolution).sum() / 2

# Calculate the energy change when flipping a spin
@numba.njit(nopython=True, nogil=True)
def calculate_energy_change(lattice: np.ndarray, x: int, y: int, L: int) -> float:
    E_i = (
        lattice[x - 1 if x > 0 else L - 1, y] +
        lattice[x + 1 if x < L - 1 else 0, y] +
        lattice[x, y - 1 if y > 0 else L - 1] +
        lattice[x, y + 1 if y < L - 1 else 0]
    )

    return 2 * lattice[x, y] * E_i

# Perform the Wolff algorithm to simulate the Ising model
@numba.njit(nopython=True, nogil=True)
def wolff_algorithm(L: int, init_up_rate: float, N: int, T: float) -> tuple:
    lattice, magnetization, energy = initialize_lattice(L, init_up_rate)
    energies = np.zeros(N, dtype=np.float64)
    magnetizations = np.zeros(N, dtype=np.float64)

    for i in range(N):
        x, y = np.random.randint(0, L, size=2)
        s_i = lattice[x, y]
        flip = [[x, y]]
        p = 1. - np.exp(-2 / T)
        f_old = [[x, y]]

        while f_old:
            f_new = []
            for x, y in f_old:
                neighbors = [
                    [x - 1 if x > 0 else L - 1, y], 
                    [x + 1 if x < L - 1 else 0, y], 
                    [x, y - 1 if y > 0 else L - 1], 
                    [x, y + 1 if y < L - 1 else 0]
                ]
                for near in neighbors:
                    if lattice[near[0], near[1]] == s_i and near not in flip:
                        if np.random.rand() < p:
                            f_new.append(near)
                            flip.append(near)
            f_old = f_new

        for x, y in flip:
            dE = calculate_energy_change(lattice, x, y, L)
            lattice[x, y] = -s_i
            energy += dE
            magnetization += -2 * s_i

        energies[i] = energy
        magnetizations[i] = magnetization

    return energies, magnetizations

# Remove initial non-equilibrium data (burn-in) based on 1-sigma convergence
@numba.njit(nopython=True, nogil=True)
def burn_in(data: np.ndarray, sigma: float = 1.0) -> np.ndarray:
    mask = np.abs(data - np.mean(data)) < sigma * np.std(data)
    if np.any(mask):
        return data[np.argmax(mask):]
    return data

# Compute statistics across multiple simulations
def compute_statistics(temp_range: np.ndarray, L_range: np.ndarray, N: int, init_up_rate: float = 0.5, simulation_func=wolff_algorithm) -> dict:
    D = 0
    for L in L_range:
        for T in temp_range:
            energies, magnetizations = simulation_func(L, init_up_rate, N, T)
            energy, hist_E, hist_M, hist_M2, hist_M4 = compute_histogram(energies, magnetizations)
            
            with open("data/hist{}.csv".format(D), "w") as out:
                out.writelines("T,{},Nsite,{}\nenergy,hist_E,hist_M,hist_M2,hist_M4\n".format(T, L**2))
                for i in range(len(energy)):
                    out.writelines("{},{:11.8e},{:11.8e},{:11.8e},{:11.8e}\n".format(energy[i], hist_E[i], hist_M[i], hist_M2[i], hist_M4[i]))
            D += 1
    return 0


def compute_histogram(energies: np.ndarray, magnetizations: np.ndarray) -> tuple:
    energies, magnetizations = burn_in(energies), burn_in(magnetizations)
    l = min(len(energies), len(magnetizations))
    energies, magnetizations = energies[-l:], abs(magnetizations[-l:])

    histogram = {}
    for e, m in zip(energies, magnetizations):
        if e in histogram:
            histogram[e] = histogram[e] + [m]
        else:
            histogram[e] = [m]
    histogram = dict(sorted(histogram.items()))

    energy_levels = list(histogram.keys())
    hist_E, hist_M, hist_M2, hist_M4 = np.zeros_like(energy_levels), np.zeros_like(energy_levels), np.zeros_like(energy_levels), np.zeros_like(energy_levels)
    for i in range(len(energy_levels)):
        Ms = np.array(histogram[energy_levels[i]])
        hist_E[i], hist_M[i], hist_M2[i], hist_M4[i] = len(Ms), np.mean(Ms), np.mean(Ms**2), np.mean(Ms**4)
    return energy_levels, hist_E, hist_M, hist_M2, hist_M4

def LinearRegression(x,y):
    x, y = np.array(x), np.array(y)
    a = sum((x-np.mean(x))*(y-np.mean(y)))/sum((x-np.mean(x))**2)
    b = np.mean(y) - a * np.mean(x)
    return a, b

File: ising.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

#define MAX_D 100
#define MAX_E 1000

#define INIT_L 2
#define FINAL_L 10
#define STEP_L 1

#define T0 2.7
#define DT_SCAN 1.0
#define DT_STEP 0.5

#define N 1000

#define UP_RATE 0.5

typedef struct {
    double energy;
    double *magnetizations;
    int count;
} Histogram;

double energy[MAX_E], hist_E[MAX_E], hist_M[MAX_E], hist_M2[MAX_E], hist_M4[MAX_E];

// Function prototypes
int cmp(const void *a, const void *b);
void initialize_lattice(int L, int** lattice, int* magnetization, double* energy);
double compute_total_energy(int** lattice, int L);
double calculate_energy_change(int** lattice, int x, int y, int L);
void wolff_algorithm(int L, double T, double* energies, double* magnetizations);
int compute_histogram(double* energies, double* magnetizations);
void compute_statistics(const double* temp_range, int temp_size, const int* L_range, int L_size);
void write_csv(const char* filename, double T, int L, double* energy, double* hist_E, double* hist_M, double* hist_M2, double* hist_M4, int size);

int main() {
    int temp_size = (int)((2 * DT_SCAN) / DT_STEP) + 1;
    double* T_range = (double*)malloc(temp_size * sizeof(double));
    int t;
    for (t = 0; t < temp_size; t++) {
        T_range[t] = T0 - DT_SCAN + t * DT_STEP;
    }

    int L_size = (FINAL_L - INIT_L) / STEP_L + 1;
    int* L_range = (int*)malloc(L_size * sizeof(int));
    int l;
    for (l = 0; l < L_size; l++) {
        L_range[l] = INIT_L + l * STEP_L;
    }

    compute_statistics(T_range, temp_size, L_range, L_size);

    free(T_range);
    free(L_range);

    return 0;
}

void compute_statistics(const double* temp_range, int temp_size, const int* L_range, int L_size) {
    int D = 0;
    int l, t;
    for (l = 0; l < L_size; l++) {
        int L = L_range[l];
        for (t = 0; t < temp_size; t++) {
            double T = temp_range[t];

            printf("#Writing file data/hist%d.csv\n", D);

            double* energies = (double*)malloc(N * sizeof(double));
            double* magnetizations = (double*)malloc(N * sizeof(double));

            wolff_algorithm(L, T, energies, magnetizations);

            int hist_count = compute_histogram(energies, magnetizations);

            char filename[50];
            sprintf(filename, "data/hist%d.csv", D);
            write_csv(filename, T, L, energy, hist_E, hist_M, hist_M2, hist_M4, hist_count);

            free(energies);
            free(magnetizations);

            printf("#T1 = %.1f  Nsite = %d\n", T, L * L);
            D++;
        }
    }
}

int cmp(const void *a, const void *b) {
    double diff = ((Histogram *)a)->energy - ((Histogram *)b)->energy;
    return (diff > 0) - (diff < 0);
}

int compute_histogram(double* energies, double* magnetizations) {
    int i, j;
    for (i = 0; i < N; i++) {
        magnetizations[i] = fabs(magnetizations[i]);
    }

    Histogram histogram[MAX_E];
    int hist_count = 0;

    for (i = 0; i < N; i++) {
        double e = energies[i];
        double m = magnetizations[i];

        int found = 0;

        for (j = 0; j < hist_count; j++) {
            if (histogram[j].energy == e) {
                histogram[j].magnetizations[histogram[j].count++] = m;
                found = 1;
                break;
            }
        }

        if (!found) {
            histogram[hist_count].energy = e;
            histogram[hist_count].magnetizations = malloc(N * sizeof(double));
            histogram[hist_count].magnetizations[0] = m;
            histogram[hist_count].count = 1;
            hist_count++;
        }
    }

    qsort(histogram, hist_count, sizeof(Histogram), cmp);

    for (i = 0; i < hist_count; i++) {
        double sum = 0.0, sum_sq = 0.0, sum_quad = 0.0;
        int count = histogram[i].count;
        for (j = 0; j < count; j++) {
            double m = histogram[i].magnetizations[j];
            sum += m;
            sum_sq += m * m;
            sum_quad += m * m * m * m;
        }
        energy[i] = histogram[i].energy;
        hist_E[i] = count;
        hist_M[i] = sum / count;
        hist_M2[i] = sum_sq / count;
        hist_M4[i] = sum_quad / count;
    }

    return hist_count;
}

void initialize_lattice(int L, int** lattice, int* magnetization, double* energy) {
    srand(time(NULL));
    *magnetization = 0;

    int i, j;
    for (i = 0; i < L; i++) {
        for (j = 0; j < L; j++) {
            double random_value = (double)rand() / RAND_MAX;
            lattice[i][j] = (random_value < UP_RATE) ? 1 : -1;
            *magnetization += lattice[i][j];
        }
    }
    *energy = compute_total_energy(lattice, L);
}

double compute_total_energy(int** lattice, int L) {
    double total_energy = 0.0;

    int i, j;
    for (i = 0; i < L; i++) {
        for (j = 0; j < L; j++) {
            int right = lattice[i][(j + 1) % L];
            int down = lattice[(i + 1) % L][j];
            total_energy -= lattice[i][j] * (right + down);
        }
    }

    return total_energy;
}

double calculate_energy_change(int** lattice, int x, int y, int L) {
    int left = lattice[x][(y - 1 + L) % L];
    int right = lattice[x][(y + 1) % L];
    int up = lattice[(x - 1 + L) % L][y];
    int down = lattice[(x + 1) % L][y];

    int neighbor_sum = left + right + up + down;
    return 2 * lattice[x][y] * neighbor_sum;
}

void wolff_algorithm(int L, double T, double* energies, double* magnetizations) {
    int** lattice = (int**)malloc(L * sizeof(int*));
    int i;
    for (i = 0; i < L; i++) {
        lattice[i] = (int*)malloc(L * sizeof(int));
    }

    int magnetization;
    double energy;
    initialize_lattice(L, lattice, &magnetization, &energy);

    double p = 1.0 - exp(-2.0 / T);

    int step;
    for (step = 0; step < N; step++) {
        int x = rand() % L;
        int y = rand() % L;

        int s_i = lattice[x][y];
        int** stack = (int**)malloc(L * L * sizeof(int*));
        int stack_size = 0;

        stack[stack_size] = (int*)malloc(2 * sizeof(int));
        stack[stack_size][0] = x;
        stack[stack_size][1] = y;
        stack_size++;

        while (stack_size > 0) {
            int* current = stack[--stack_size];
            int cx = current[0];
            int cy = current[1];
            free(current);

            if (lattice[cx][cy] == s_i) {

                int neighbors[4][2] = {
                    {cx, (cy - 1 + L) % L},
                    {cx, (cy + 1) % L},
                    {(cx - 1 + L) % L, cy},
                    {(cx + 1) % L, cy}
                };

                for (i = 0; i < 4; i++) {
                    int nx = neighbors[i][0];
                    int ny = neighbors[i][1];
                    if (lattice[nx][ny] == s_i && ((double)rand() / RAND_MAX) < p) {
                        
                        energy += calculate_energy_change(lattice, cx, cy, L);
                        lattice[cx][cy] = -s_i;
                        magnetization += -2 * s_i;

                        stack[stack_size] = (int*)malloc(2 * sizeof(int));
                        stack[stack_size][0] = nx;
                        stack[stack_size][1] = ny;
                        stack_size++;
                    }
                }
            }
        }
        
        for (i = 0; i < stack_size; i++) {
            free(stack[i]);
        }
        free(stack);

        energies[step] = energy;
        magnetizations[step] = magnetization;
    }

    for (i = 0; i < L; i++) {
        free(lattice[i]);
    }
    free(lattice);
}

void write_csv(const char* filename, double T, int L, double* energy, double* hist_E, double* hist_M, double* hist_M2, double* hist_M4, int size) {
    FILE* file = fopen(filename, "w");
    if (!file) {
        perror("Failed to open file");
        exit(1);
    }

    fprintf(file, "T,%.1f,Nsite,%d\n", T, L * L);
    fprintf(file, "energy,hist_E,hist_M,hist_M2,hist_M4\n");

    int i;
    for (i = 0; i < size; i++) {
        fprintf(file, "%.1f,%e,%e,%e,%e\n", energy[i], hist_E[i], hist_M[i], hist_M2[i], hist_M4[i]);
    }

    fclose(file);
}
